### DISCUSSION TITLE
🚀 Introducing QQN: A hyperparameter-free, JAX-native Quadratic Quasi-Newton Optimizer

### DISCUSSION THREAD

**Discussion by acharneski** [jax-ml/jax#38701] [Show and tell]
Hi everyone,

I'm excited to share [**`qqn-jax`**](https://pypi.org/project/qqn-jax/0.1.0/), a new pure-JAX/Optax optimizer designed for ill-conditioned, anisotropic landscapes where first-order methods (like Adam/SGD) stall out and standard second-order methods (like L-BFGS) become brittle.

If you’ve ever been frustrated by tuning learning rates on full-batch physics simulations, complex geometric problems, or smooth/ill-curved objectives, QQN was built for you.

### 💡 The Core Idea: The Quadratic Path
Standard L-BFGS proposes a confident second-order step, and if it overshoots, it backtracks along a straight line. This throws away a structural fact: at small step sizes, we *know* the best local direction is the negative gradient.

QQN fixes this by replacing the straight-line search with a **quadratic path** that blends steepest descent with a quasi-Newton oracle (L-BFGS by default):

```text
d(t) = t(1 - t)(-∇f) + t²(-H∇f),   t ∈ [0, 1]
```

* At `t = 0`, the path's tangent is pure gradient descent.
* At `t = 1`, it arrives exactly at the pure L-BFGS prediction.
* A single line search traverses this curve, picking the interpolation parameter `t` and step size together.

**The result? There is no learning rate to tune.** The line search automatically discovers the right blend of conservative first-order and aggressive second-order behavior at every single iteration. Because the path is anchored by the gradient at `t=0`, it guarantees descent *regardless of how poor the L-BFGS approximation is*.

### 🛠 Built for the JAX Ecosystem
I built QQN from the ground up to play perfectly with JAX's functional model:
* **Fully Transformable:** The entire solver loop is written using `jax.lax.while_loop`. A full optimization run is a single traceable operation. You can `jax.jit` it for XLA dispatch, `jax.vmap` it to solve massive batches of initializations simultaneously, or even `jax.grad` through the solver.
* **Optax Under the Hood:** Delegates L-BFGS scaling and zoom line searches to Optax, keeping the ecosystem tight.
* **Highly Modular:** QQN is actually a combiner of four orthogonal axes: **Gradient**, **Oracle** (swap L-BFGS for Momentum, Shampoo, or Anderson), **Search** (Armijo, Strong Wolfe, or our custom cubic-Hermite spline refinement), and **Region** (Box, Orthant, Trust-Region constraints).

### 📊 When should you use it?
To be clear: **QQN is not a drop-in replacement for Adam on large-scale stochastic minibatch training.** Adam remains king for noisy, high-dimensional deep learning.

QQN shines on **smooth, full-batch, or deterministic objectives** where curvature structure matters. In our benchmarks (e.g., a 335k parameter MLP on full-batch Fashion-MNIST), QQN completely dominates standard L-BFGS on the Pareto frontier. It requires fewer iterations to hit tight loss targets *and* is cheaper per iteration (16ms/it vs 20ms/it) because our curved path allows us to use a bare Armijo search instead of an expensive Zoom search.

### 🚀 Quick Start

```bash
pip install qqn-jax
```

```python
import jax
import jax.numpy as jnp
from qqn_jax import QQN

# Define your objective
def rosenbrock(x):
    return (1 - x[0]) ** 2 + 100 * (x[1] - x[0] ** 2) ** 2

# Initialize the solver (no learning rate needed!)
solver = QQN(rosenbrock, maxiter=100, tol=1e-6)
init_params = jnp.array([-1.2, 1.0])

# JIT compile and run the entire optimization loop
run_jit = jax.jit(solver.run)
params, state = run_jit(init_params)

print(f"Converged to: {params} in {state.iter} iterations")
```

### 📖 Learn More
Extensive documentation including the theory, the algorithm's genesis, and the empirical results are available on the repo.

* **GitHub Repo:** https://github.com/SimiaCryptus/qqn-jax
* **Algorithm Deep Dive:** https://github.com/SimiaCryptus/qqn-jax/blob/master/docs/theory/algorithm.md
* **Empirical Results:** https://github.com/SimiaCryptus/qqn-jax/tree/master/results

I'd love for the JAX community to try it out on your weird, ill-conditioned, or physics-informed loss landscapes. Let us know what you think, and feel free to open issues or PRs!
